include/flamegpu/runtime/utility/DeviceMacroProperty.cuh

```diff
@@ -0,0 +1,452 @@
+#endif
+    /**
+     * Access the next dimension of the array
+     * @throws exception::DeviceError If i >= I. 
+     * @throws exception::DeviceError If template arguments I, J, K , W are all 1. Which denotes the macro property has no dimensions remaining to be indexed. 
+     */
+    __device__ __forceinline__ DeviceMacroProperty<T, J, K, W, 1> operator[](unsigned int i) const;
+    /**
+     * atomic add
+     * @param val The 2nd operand
+     * @return a reference to this
+     * Note, taking value of the returned object will fail, due to the risk of atomic conflicts
+     * @note Only suitable where T is type int32_t, uint32_t, uint64_t, float, double
+     */
+    __device__ __forceinline__ DeviceMacroProperty<T, I, J, K, W>& operator +=(const T& val);
+    /**
+     * atomic subtraction
+     * @param val The 2nd operand
+     * @return a reference to this
+     * Note, taking value of the returned object will fail, due to the risk of atomic conflicts
+     * @note Only suitable where T is type int32_t or uint32_t
+     */
+    __device__ __forceinline__ DeviceMacroProperty<T, I, J, K, W>& operator -=(const T& val);
+    /**
+     * atomic add
+     * @param val The 2nd operand
+     * @return (this + val)
+     * @note Only suitable where T is type int32_t, uint32_t, uint64_t, float, double
+     */
+    __device__ __forceinline__ T operator+(const T& val) const;
+    /**
+     * atomic subtraction
+     * @param val The 2nd operand
+     * @return (this - val)
+     * @note Only suitable where T is type int32_t or uint32_t
+     */
+    __device__ __forceinline__ T operator-(const T& val) const;
+    /**
+     * atomic pre-increment
+     * @return the value after the increment operation is performed
+     * @note A reference is not returned, as this makes it easy to fall into the trap (race condition) of reading it's value.
+     * @note Only suitable where T is type uint32_t
+     */
+    __device__ __forceinline__ T operator++();
+    /**
+     * atomic pre-decrement
+     * @return the value after the decrement operation is performed
+     * @note A reference is not returned, as this makes it easy to fall into the trap (race condition) of reading it's value.
+     * @note Only suitable where T is type uint32_t
+     */
+    __device__ __forceinline__ T operator--();
+    /**
+     * atomic post-increment
+     * @return the value before increment operation is performed
+     * @note Only suitable where T is type uint32_t
+     */
+    __device__ __forceinline__ T operator++(int);
+    /**
+     * atomic post-decrement
+     * @return the value before increment operation is performed
+     * @note Only suitable where T is type uint32_t
+     */
+    __device__ __forceinline__ T operator--(int);
+    /**
+     * atomic min
+     * @return min(this, val)
+     * @note Only suitable where T is type int32_t, uint32_t, uint64_t
+     */
+    __device__ __forceinline__ T min(T val);
```

Comment:
Again i'm not keen on how it is not immediately clear that this is atomic without reading the docs, but that might just be an artifact of using CUDA too much.

Reply:
The understanding should be that anything except reading the value is atomic in an agent fn with macro env prop.